rtp/src/main/kotlin/org/jitsi/rtp/rtcp/rtcpfb/transport_layer_fb/ccfb/RtcpFbCcfbPacket.kt

```diff
@@ -0,0 +1,330 @@
+    kNotEct(0), // Not ECN-Capable Transport
+    kEct1(1), // ECN-Capable Transport
+    kEct0(2), // Not used by L4s (or webrtc.)
+    kCe(3); // Congestion experienced
+
+    companion object {
+        fun fromBits(bits: Byte): EcnMarking {
+            entries.forEach {
+                if (it.bits == bits) {
+                    return it
+                }
+            }
+            throw IllegalArgumentException("Invalid ECNMarking value: $bits")
+        }
+    }
+}
+
+sealed class PacketInfo(
+    val ssrc: Long = 0,
+    val sequenceNumber: Int = 0
+) {
+    override fun toString(): String {
+        return "ssrc=$ssrc, sequenceNumber=$sequenceNumber"
+    }
+}
+
+class UnreceivedPacketInfo(ssrc: Long = 0, sequenceNumber: Int = 0) : PacketInfo(ssrc, sequenceNumber) {
+    override fun toString(): String {
+        return "Unreceived: " + super.toString()
+    }
+}
+
+class ReceivedPacketInfo(
+    ssrc: Long = 0,
+    sequenceNumber: Int = 0,
+    val arrivalTimeOffset: Duration = MIN_DURATION,
+    val ecn: EcnMarking = EcnMarking.kNotEct
+) : PacketInfo(ssrc, sequenceNumber) {
+    override fun toString(): String {
+        return "Received: " + super.toString() + ", arrivalTimeOffset=$arrivalTimeOffset, ecn=$ecn"
+    }
+}
+
+private const val kHeaderPerMediaSsrcLength = 8
+private const val kTimestampLength = 4
+
+class RtcpFbCcfbPacketBuilder(
+    val rtcpHeader: RtcpHeaderBuilder = RtcpHeaderBuilder(),
+    var mediaSourceSsrc: Long = -1,
+    // `Packets` MUST be sorted in sequence_number order per SSRC. There MUST not
+    // be missing sequence numbers between `Packets`. `Packets` MUST not include
+    // duplicate sequence numbers.
+    val packets: List<PacketInfo> = listOf(),
+    val reportTimestampCompactNtp: Long
+) {
+    private fun blockLength(): Int {
+        var totalSize = RtcpFbPacket.HEADER_SIZE + kTimestampLength
+
+        packets.groupBy { it.ssrc }.forEach { (ssrc, packets) ->
+            val blockSize = packets.size * 2
+            totalSize += kHeaderPerMediaSsrcLength + blockSize + RtpUtils.getNumPaddingBytes(blockSize)
+        }
+
+        return totalSize
+    }
+
+    fun writeTo(buf: ByteArray, offset: Int) {
+        val blockLength = blockLength()
+        val paddingBytes = RtpUtils.getNumPaddingBytes(blockLength)
+        val packetSize = blockLength + paddingBytes
+        require(buf.size - offset >= packetSize) {
+            "Buffer of size ${buf.size} with offset $offset too small for CCFB packet of size $packetSize"
+        }
+        rtcpHeader.apply {
+            packetType = TransportLayerRtcpFbPacket.PT
+            reportCount = RtcpFbCcfbPacket.FMT
+            length = RtpUtils.calculateRtcpLengthFieldValue(blockLength)
+        }.writeTo(buf, offset)
+
+        RtcpFbPacket.setMediaSourceSsrc(buf, offset, mediaSourceSsrc)
+        var position = offset + RtcpFbPacket.HEADER_SIZE
+
+        packets.groupBy { it.ssrc }.forEach { (ssrc, packets) ->
```

Comment:
How efficient is `groupBy`? A single `build()` would call it 3 times. Why not maintain it as a map to start with?